feat(DIA-69): add progressive onboarding model for saved artworks

araujobarret opened the pull request:
This PR resolves [DIA-69] <!-- eg [PROJECT-XXXX] -->

### Description

Adds the progressive onboarding for saved artworks.

Adds a new store model Progressive Onboarding and instructions about how to use it + short intro in `docs/add_progressive_onboarding.md`.

User logged in 

https://github.com/artsy/eigen/assets/15792853/5487679c-6067-470f-b7f4-718adadb6048

User  signing up

https://github.com/artsy/eigen/assets/15792853/c91ce88c-e5a6-4d49-912a-7319161de208

### PR Checklist

- [x] I have tested my changes on **iOS** and **Android**.
- [x] I hid my changes behind a **[feature flag]**, or they don't need one.
- [x] I have included **screenshots** or **videos**, or I have not changed the UI.
- [x] I have added **tests**, or my changes don't require any.
- [x] I added an **[app state migration]**, or my changes do not require one.
- [x] I have documented any **follow-up work** that this PR will require, or it does not require any.
- [x] I have added a **changelog entry** below, or my changes do not require one.

### To the reviewers 👀

- [ ] I would like **at least one** of the reviewers to **run** this PR on the simulator or device.

<details><summary>Changelog updates</summary>

### Changelog updates

<!-- 📝 Please fill out at least one of these sections. -->
<!-- ⓘ 'User-facing' changes will be published as release notes. -->
<!-- ⌫ Feel free to remove sections that don't apply. -->
<!-- • Write a markdown list or just a single paragraph, but stick to plain text. -->
<!-- 📖 eg. `Enable lotsByFollowedArtists - john` or `Fix phone input misalignment - mary`. -->
<!-- 🤷‍♂️ Replace this entire block with the hashtag `#nochangelog` to avoid updating the changelog. -->
<!-- ⚠️ Prefix with `[NEEDS EXTERNAL QA]` if a change requires external QA -->

#### Cross-platform user-facing changes

- adds progressive onboarding for saved artworks - araujobarret

#### iOS user-facing changes

-

#### Android user-facing changes

-

#### Dev changes

-

<!-- end_changelog_updates -->

</details>

Need help with something? Have a look at our [docs], or get in touch with us.

[app state migration]: ../blob/main/docs/adding_state_migrations.md
[feature flag]: ../blob/main/docs/developing_a_feature.md
[docs]: ../blob/main/docs/README.md


[DIA-69]: https://artsyproduct.atlassian.net/browse/DIA-69?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

araujobarret commented on src/app/Components/ArtworkLists/views/SelectArtworkListsForArtworkView/components/StickyBottomContent.tsx:
```diff
@@ -36,6 +38,7 @@ export const StickyBottomContent: FC<BottomSheetFooterProps> = ({ animatedFooter
     } else {
       dismiss(ArtworkListsViewName.SelectArtworkListsForArtwork)
     }
+    updateProfileTab()
```
once collector save the artwork we force the profile tab store prop to display the next popover

olerichter00 replied:
Maybe we could give this function a name that indicates its purpose, to show the next onboarding popover.

araujobarret replied:
changed in https://github.com/artsy/eigen/pull/9224/commits/608a46458e2ec55d25f7bebc56df62025a7cf114

araujobarret commented on src/app/Components/ProgressiveOnboarding/ProgressiveOnboardingSaveArtwork.tsx:
```diff
@@ -0,0 +1,59 @@
+      return <>{children}</>
+    }
+
+    return <ElementInView onVisible={() => setIsInView(true)}>{children}</ElementInView>
```
There is no guarantee that the Artwork will be always on top, sometimes the grid is down there, this `ElementInView` helps to control this aspect.

ArtsyOpenSource commented:
<!--
  0 failure: 
  0 warning: 
  
  1 markdown notices
  DangerID: danger-id-Danger;
-->



### This PR contains the following changes:

- Cross-platform user-facing changes (adds progressive onboarding for saved artworks - araujobarret)
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against f533247cf58b2c0fe984a219c81563b95a2e5db6
</p>

araujobarret commented on src/app/Scenes/ArtworkLists/ArtworkLists.tsx:
```diff
@@ -93,11 +94,15 @@ export const ArtworkLists = () => {
-  </Suspense>
-)
+export const ArtworkListsQR = () => {
+  useDismissSavedHighlight()
```
This is only used in the Profile > Saved works tab, so no harm calling the hook from here

araujobarret commented on src/app/Scenes/Home/Home.tsx:
```diff
@@ -130,6 +131,9 @@ export interface HomeProps extends ViewProps {
 }
 
 const Home = memo((props: HomeProps) => {
+  useDismissSavedArtwork(
```
we need to do this in the home, is the best place to check if the user saved any artwork before to avoid calling the popover on the wrong screen.

araujobarret commented on src/app/store/migration.ts:
```diff
@@ -291,6 +292,16 @@ export const artsyAppMigrations: Migrations = {
       pushPermissionsRequestedThisSession: false,
     }
   },
+  [Versions.AddProgressiveOnboardingModel]: (state) => {
```
as discussed with @olerichter00 the Progressive Onboarding API is simpler than the visual clue, so we're starting the deprecation process of visual clue in favor of Progressive onboarding.

damassi replied:
Before merging, can we add a full TLDR code example in the PR description showing how to use this?

olerichter00 commented on src/app/Components/ProgressiveOnboarding/ProgressiveOnboardingFindSavedArtwork.tsx:
```diff
@@ -0,0 +1,46 @@
+    (state) => state.bottomTabs.sessionState.tabProps.profile
+  )
+
+  const isDismissed = _isDismissed("find-saved-artwork").status
+
+  const isFindSavedArtworkDisplayable = !isDismissed && !!profileTabProps?.savedArtwork
```
I find the naming around `_isDismissed` a bit confusing. Maybe we could avoid the renaming like this?

```suggestion
  const isFindSavedArtworkDisplayable = !isDismissed("find-saved-artwork").status && !!profileTabProps?.savedArtwork
```

araujobarret replied:
makes sense 👍 
Added the changes here https://github.com/artsy/eigen/pull/9224/commits/a8798b0fe53c84dd9fee41b0adc347d1c8fa4874 because was missing one change from the suggestion

damassi commented on src/app/Components/ArtworkGrids/ArtworkGridItem.tsx:
```diff
@@ -363,6 +364,15 @@ export const Artwork: React.FC<ArtworkProps> = ({
                         width={SAVE_ICON_SIZE}
                         fill="blue100"
                       />
+                    ) : itemIndex === 0 ? (
```
Can we avoid double ternaries? Is there another way to write this to make it clearer?

araujobarret replied:
I don't think there's a simpler solution to avoid the ternar. The popover needs to wrap the icon, and the component has some relay query so we mustn't wrap all the items in a grid, that is not so performant.

The other solution would be an intermediate component that receives the index and wraps or not with the element but would increase the complexity just for ternary's sake.

damassi replied:
maybe something like:
```tsx
<>
  {(() => {
    if (thing) {
      return <Foo />
    }
    if (otherThing) {
      return <Bar />
    }
    if (otherThingThree) {
      return <Baz />
    }
  })()}
</>
```

damassi replied:
(Not sure if this is better... just thinking)

damassi commented on src/app/Scenes/MyProfile/MyProfileHeaderMyCollectionAndSavedWorks.tsx:
```diff
@@ -43,6 +44,14 @@ export const MyProfileHeaderMyCollectionAndSavedWorks: React.FC<Props> = ({ init
       },
     })
   }
+  if (isDismissed("find-saved-artwork").status && !isDismissed("save-highlight").status) {
```
Maybe add a comment above this line explaining a bit?

damassi reviewed (dismissed):
Looks really good @araujobarret - only request is that we 
- update the PR description with an example of how to use (too much code in PR to fully understand at a glance); and 
- maybe add some other docs since the old indicator is being deprecated

araujobarret commented:
Thanks for the inputs @damassi !
I've addressed all of them https://github.com/artsy/eigen/pull/9224/commits/0aff7c0f073cb4ab7bc3ee55a801a0a30d875752, created a doc explaining how to use the model with some examples and added deprecation x-reference in the visual clue docs(updated the description with the doc reference too) 👍 
Please let me know if you think we need something else.

damassi reviewed (approved):
Docs and changes look great @araujobarret 💯
